# Day 3 - Data Exploration, Part 2: Basic RNA-seq Plots
## Friday, October 23rd, 2020


<br>

### Join the Slack channels! 
#### https://join.slack.com/t/cvrcbioinform-zmq7258/shared_invite/zt-hja1edwl-RPS1PxgXixBjfUStcW_JTg

<br>  

##

In this session we will focus on plotting the data that we cleaned up and normalized in the previous session ([Part 1, Data Manipulation](https://github.com/florschlamp/CVRC_NYU_Langone/tree/master/Bioinformatics_Workshop_Series_2020/Day_2)). You will learn to use R to do pairwise scatterplots to compare replicates and samples. You will learn the basics of the pheatmap R package to make heatmaps of the similarity between samples, and we will introduce the basics of ggplot R package to plot individual genes of interest. These include dot plots to represent gene expression per sample and per group, barplots, and boxplots. You will also learn how to plot multiple genes at the same time or genes in paired data and time series.


<br>

### REQUIREMENTS TO PARTICIPATE IN THIS SESSION:
 
Make sure you have the following packages installed in Rstudio: ggplot2, pheatmap, RColorBrewer, and reshape2.  
Example of how to install R packages:
```r
# installing ggplot2: 
install.packages("ggplot2")

# if the installation worked, you should be able to run the following line without error messages:
library(ggplot2)
```

### OFFICE HOURS:
 
Friday Oct. 23rd
* 11am to noon  
* 1:30pm until the start of the session at 2:30pm  
 
Zoom link:
https://nyulangone.zoom.us/j/98456141720
